### PR TITLE
add nosnakecase lint

### PR DESCRIPTION
**Pull request by sivchari**
nosnakecase is a linter that detects snake case of variable naming and function name.
https://github.com/sivchari/nosnakecase

### CODE REVIEW

**Comment by ldez** [comment]
In order for a pull request adding a linter to be reviewed, the linter and the PR must follow some requirements.

### Pull Request Description

- [ ] It must have a link to the linter repository.
- [ ] It must provide a short description about the linter.

### Linter

- [ ] It must not be a duplicate of another linter or a rule of a linter. (the team will help to verify that)
- [ ] It must have a valid license and the file must contain the required information by the license, ex: author, year, etc.
- [ ] It must use [`go/analysis`](https://golangci-lint.run/contributing/new-linters/).
- [ ] It must have a valid tag, ex: `v1.0.0`, `v0.1.0`.
- [ ] It must not contain `init()`.
- [ ] It must not contain `panic()`, `log.fatal()`, `os.exit()`, or similar.
- [ ] It must not have false positives/negatives. (the team will help to verify that)
- [ ] It must have tests inside golangci-lint.

### The Linter Tests Inside Golangci-lint

- [ ] They must have at least one std lib import.
- [ ] They must work with `T=<name of the linter test file>.go make test_linters`. (the team will help to verify that)

### `.golangci.example.yml`

- [ ] The linter must be added to the list of available linters (alphabetical case-insensitive order).
  - `enable` and `disable` options
- [ ] If the linter has a configuration, the exhaustive configuration of the linter must be added (alphabetical case-insensitive order)
  - The values must be different from the default ones.
  - The default values must be defined in a comment.
  - The option must have a short description.

### Others Requirements

- [ ] The files (tests and linter) inside golangci-lint must have the same name as the linter.
- [ ] The `.golangci.yml` of golangci-lint itself must not be edited and the linter must not be added to this file.
- [ ] The linters must be sorted in the alphabetical order (case-insensitive) in the `Manager.GetAllSupportedLinterConfigs(...)` and `.golangci.example.yml`.
- [ ] The load mode (`WithLoadMode(...)`):
  - if the linter doesn't use types: `goanalysis.LoadModeSyntax`
  - `goanalysis.LoadModeTypesInfo` required `WithLoadForGoAnalysis()` in the `Manager.GetAllSupportedLinterConfigs(...)`
- [ ] The version in `WithSince(...)` must be the next minor version (`v1.X.0`) of golangci-lint.

### Recommendations

- [ ] The linter should not use SSA. (currently, SSA does not support generics)
- [ ] The linter repository should have a CI, tests, a readme and linting.
- [ ] The linter should be published as a binary. (useful to diagnose bug origins)

---

The golangci-lint team will edit this comment to check the boxes before and during the review.

**This checklist does not imply that we will accept the linter.**

**Comment by sivchari** [comment]
I fixed it and all checks have passed.

**Comment by ldez** [comment]
You say:
> nosnakecase is a linter that detects snake case of variable naming and function name.

but the linter detects:
- some constants
- named returns

So the description doesn't seem clear about the real scope of this linter.

Also, the linter doesn't detect global variables, global constants, interface names, struct names, struct field name, function in struct field, interface method names, etc.

Otherwise, I recommend adding this link to your linter `readme.md` file: https://go.dev/doc/effective_go#mixed-caps 

As you can see in the following sample, there are non-detected elements:

<details>

```go
package sandbox

// global variable name with underscore.
var v_v = 0 // FAIL: not detected

// global constant name with underscore.
const c_c = 0 // FAIL: not detected

// struct name with underscore.
type S_a struct { // FAIL: not detected
	fi int
}

// non-exported struct field name with underscore.
type Sa struct {
	fi_a int // FAIL: not detected
}

// function as struct field, with parameter name with underscore.
type Sb struct {
	fib func(p_a int) // FAIL: not detected
}

// exported struct field with underscore.
type Sc struct {
	Fi_A int // FAIL: not detected
}

// function as struct field, with return name with underscore.
type Sd struct {
	fib func(p int) (r_a int) // FAIL: not detected
}

// interface name with underscore.
type I_a interface { // FAIL: not detected
	fn(p int)
}

// interface with parameter name with underscore.
type Ia interface {
	fn(p_a int) // FAIL: not detected
}

// interface with parameter name with underscore.
type Ib interface {
	Fn(p_a int) // FAIL: not detected
}

// function as struct field, with return name with underscore.
type Ic interface {
	Fn_a() // FAIL: not detected
}

// interface with return name with underscore.
type Id interface {
	Fn() (r_a int) // FAIL: not detected
}

// function name with underscore.
func f_a() {}

// function's parameter name with underscore.
func fb(p_a int) {}

// named return with underscore.
func fc() (r_b int) {
	return 0
}

// local variable (short declaration) with underscore.
func fd(p int) int {
	v_b := p * 2

	return v_b
}

// local constant with underscore.
func fe(p int) int {
	const v_b = 2

	return v_b * p
}

// local variable with underscore.
func ff(p int) int {
	var v_b = 2

	return v_b * p
}

// inner function, parameter name with underscore.
func fg() {
	fgl := func(p_a int) {} // FAIL: not detected
	fgl(1)
}

type Foo struct{}

// method name with underscore.
func (f Foo) f_a() {}

// method's parameter name with underscore.
func (f Foo) fb(p_a int) {}

// named return with underscore.
func (f Foo) fc() (r_b int) { return 0 }

// local variable (short declaration) with underscore.
func (f Foo) fd(p int) int {
	v_b := p * 2

	return v_b
}

// local constant with underscore.
func (f Foo) fe(p int) int {
	const v_b = 2

	return v_b * p
}

// local variable with underscore.
func (f Foo) ff(p int) int {
	var v_b = 2

	return v_b * p
}

func fna(a, p_a int) {}

func fna1(a string, p_a int) {}

func fnb(a, b, p_a int) {}

func fnb1(a, b string, p_a int) {}

func fnc(p_a int, p_b int) {}

func fnd(p_a int, p_b int, p_c int) {}

```

</details>


The item "They must have at least one std lib import." is still not implemented.

**Comment by sivchari** [comment]
Hi, @ldez. Sorry to late.
This linter can't detect named imports, too. So I'll fixed it. Thanks.

**Comment by sivchari** [comment]
Hi @ldez , I fixed it.

File: test/testdata/cgo/main.go [hunk 1]
```diff
@@ -14,6 +14,7 @@ import (
 	"unsafe"
 )
 
+//nolint:nosnakecase
```
Comment (ldez) [inline test/testdata/cgo/main.go:17]:
```suggestion
```

**Review by ldez — CHANGES_REQUESTED** [review]

File: test/testdata/cgo/main.go (hunk 1 above)
Comment (sivchari) [reply test/testdata/cgo/main.go:17]: If the directive is deleted, CI  is failed as the following reason.

```console
--- PASS: TestTestsAreLintedByDefault (2.52s)
=== RUN   TestCgoOk
level=info msg="[test] stderr: "
level=info msg="[test] ran [../golangci-lint run --go=1.17 --internal-cmd-test --allow-parallel-runners --no-config --enable-all testdata/cgo] in 3.078396167s"
    testshared.go:53: 
        	Error Trace:	testshared.go:53
        	            				run_test.go:98
        	Error:      	Not equal: 
        	            	expected: ""
Error:         	            	actual  : "testdata/cgo/main.go:18:8: _Cfunc_CString is used under score. You should use mixedCap or MixedCap. (nosnakecase)\n\tcs := C.CString(\"Hello from stdio\\n\")\n\t      ^\ntestdata/cgo/main.go:19:2: _Cfunc_myprint is used under score. You should use mixedCap or MixedCap. (nosnakecase)\n\tC.myprint(cs)\n\t^\ntestdata/cgo/main.go:20:11: _cgo0 is used under score. You should use mixedCap or MixedCap. (nosnakecase)\n\tC.free(unsafe.Pointer(cs))\n\t         ^\n"
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1,10 @@
Error:         	            	+testdata/cgo/main.go:18:8: _Cfunc_CString is used under score. You should use mixedCap or MixedCap. (nosnakecase)
        	            	+	cs := C.CString("Hello from stdio\n")
        	            	+	      ^
Error:         	            	+testdata/cgo/main.go:19:2: _Cfunc_myprint is used under score. You should use mixedCap or MixedCap. (nosnakecase)
        	            	+	C.myprint(cs)
        	            	+	^
Error:         	            	+testdata/cgo/main.go:20:11: _cgo0 is used under score. You should use mixedCap or MixedCap. (nosnakecase)
        	            	+	C.free(unsafe.Pointer(cs))
        	            	+	         ^
        	Test:       	TestCgoOk
        	Messages:   	exit code is 1
    testshared.go:54: 
        	Error Trace:	testshared.go:54
        	            				run_test.go:98
        	Error:      	Not equal: 
        	            	expected: 0
        	            	actual  : 1
        	Test:       	TestCgoOk
Error:         	Messages:   	output is testdata/cgo/main.go:18:8: _Cfunc_CString is used under score. You should use mixedCap or MixedCap. (nosnakecase)
        	            		cs := C.CString("Hello from stdio\n")
        	            		      ^
Error:         	            	testdata/cgo/main.go:19:2: _Cfunc_myprint is used under score. You should use mixedCap or MixedCap. (nosnakecase)
        	            		C.myprint(cs)
        	            		^
Error:         	            	testdata/cgo/main.go:20:11: _cgo0 is used under score. You should use mixedCap or MixedCap. (nosnakecase)
        	            		C.free(unsafe.Pointer(cs))
```

**Comment by sivchari** [comment]
Sorry, I fixed a false positive of following.

```
package a_test // false positive
```

**Comment by ldez** [comment]
I will open a PR on your linter.

**Comment by sivchari** [comment]
All tests are passed 👍

**Review by ldez — APPROVED** [review]
LGTM

**Pull request merged**